hail/python/benchmark/hail/conftest.py

```diff
@@ -0,0 +1,313 @@
+import json
+import logging
+import os
+import platform
+import sys
+from contextlib import contextmanager
+from datetime import datetime, timezone
+from typing import Any, Dict, List, Literal
+
+import pytest
+from _pytest.runner import pytest_runtest_protocol as runner_runtest_protocol
+
+import hail as hl
+from benchmark.hail.fixtures import (
+    HG00187,
+    HG00190,
+    HG00308,
+    HG00313,
+    HG00320,
+    HG00323,
+    HG00339,
+    HG00373,
+    HG00524,
+    HG00553,
+    HG00590,
+    HG00592,
+    HG00637,
+    HG01058,
+    HG01356,
+    HG01357,
+    HG01377,
+    HG01378,
+    HG01383,
+    HG01384,
+    HG02188,
+    HG02223,
+    HG02224,
+    HG02230,
+    HG03088,
+    HG03363,
+    HG03578,
+    HG03805,
+    HG03833,
+    HG04001,
+    HG04158,
+    NA11830,
+    NA12249,
+    NA18507,
+    NA18530,
+    NA18534,
+    NA18609,
+    NA18613,
+    NA18618,
+    NA18969,
+    NA19456,
+    NA19747,
+    NA20317,
+    NA20760,
+    NA20769,
+    NA20796,
+    NA20802,
+    NA21099,
+    NA21122,
+    NA21123,
+    balding_nichols_5k_5k,
+    chr22_gvcfs,
+    empty_gvcf,
+    gnomad_dp_sim,
+    many_ints_ht,
+    many_ints_tsv,
+    many_partitions_ht,
+    many_strings_ht,
+    many_strings_tsv,
+    onekg_chr22,
+    profile25_mt,
+    profile25_vcf,
+    random_doubles_mt,
+    random_doubles_tsv,
+    resource_dir,
+    sim_ukb_bgen,
+    sim_ukb_sample,
+    single_gvcf,
+)
+from benchmark.hail.utils import (
+    get_peak_task_memory,
+    init_hail_for_benchmarks,
+    run_with_timeout,
+    select,
+)
+from hail.utils.java import Env
+
+
+@contextmanager
+def init_hail(run_config):
+    init_hail_for_benchmarks(run_config)
+    try:
+        yield
+    finally:
+        hl.stop()
+
+
+results = pytest.StashKey[Dict[str, Dict[str, Any]]]()
+
+# item stash
+start = pytest.StashKey[datetime]()
+end = pytest.StashKey[datetime]()
+iteration = pytest.StashKey[int]()
+runs = pytest.StashKey[List[Dict[str, Any]]]()
+
+# used internally
+context = pytest.StashKey[Literal['burn_in', 'benchmark']]()
+
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_runtest_protocol(item, nextitem):
+    run_config = item.session.config.run_config
+    # Initialise hail before running every benchmark for two reasons:
+    # - each benchmark runs in a clean hail session
+    # - our means of getting max task memory is quite crude (regex on logs)
+    #   and a fresh session provides a new log
+    with init_hail(run_config):
+        if run_config.iterations is not None:
+            burn_in_iterations = 0
```

Comment:
Why does overriding iterations affect burn_in_iterations?

Reply:
I don't have a use case for overriding the number of burn in iterations globally.
My use case for globally overriding total iterations is for our future work in analysing the benchmarks (following https://www.zora.uzh.ch/id/eprint/170445/1/emse_smb_cloud.pdf). Here, we don't know how many burn in iterations we want to run yet and so we'll just run a bunch and follow their methodology to compute number or burn in iterations, timed iterations and jobs.